pkg/syncer/admissionhandler/admissionhandler.go

```diff
@@ -0,0 +1,207 @@
+		return
+	}
+
+	var admissionResponse *admissionv1.AdmissionResponse
+	ar := admissionv1.AdmissionReview{}
+	codecs := serializer.NewCodecFactory(runtime.NewScheme())
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		log.Errorf("Can't decode body: %v", err)
+		admissionResponse = &admissionv1.AdmissionResponse{
+			Result: &metav1.Status{
+				Message: err.Error(),
+			},
+		}
+	} else {
+		if r.URL.Path == "/validate-storageclass" {
+			log.Debugf("request URL path is /validate-storageclass")
+			log.Debugf("admissionReview: %+v", ar)
+			admissionResponse = validateStorageClass(ctx, &ar)
+			log.Debugf("admissionResponse: %+v", admissionResponse)
+		}
+	}
+
+	admissionReview := admissionv1.AdmissionReview{}
+	admissionReview.APIVersion = "admission.k8s.io/v1"
+	admissionReview.Kind = "AdmissionReview"
+	if admissionResponse != nil {
+		admissionReview.Response = admissionResponse
+		if ar.Request != nil {
+			admissionReview.Response.UID = ar.Request.UID
+		}
+	}
+	resp, err := json.Marshal(admissionReview)
+	if err != nil {
+		log.Errorf("Can't encode response: %v", err)
+		http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+	}
+	log.Debugf("Ready to write response")
+	if _, err := w.Write(resp); err != nil {
+		log.Errorf("Can't write response: %v", err)
+		http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+	}
+}
+
+// validateStorageClass helps validate AdmissionReview requests for StroageClass
+func validateStorageClass(ctx context.Context, ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	log := logger.GetLogger(ctx)
+	req := ar.Request
+
+	var result *metav1.Status
+	allowed := true
+
+	switch req.Kind.Kind {
+	case "StorageClass":
+		sc := stroagev1.StorageClass{}
+		log.Debugf("JSON req.Object.Raw: %v", string(req.Object.Raw))
+		if err := json.Unmarshal(req.Object.Raw, &sc); err != nil {
+			log.Error("error deserializing storage class")
+			return &admissionv1.AdmissionResponse{
+				Result: &metav1.Status{
+					Message: err.Error(),
+				},
+			}
+		}
+		log.Infof("Validating StorageClass: %q", sc.Name)
+		// AllowVolumeExpansion check for kubernetes.io/vsphere-volume provisioner
+		if sc.Provisioner == "kubernetes.io/vsphere-volume" {
+			if sc.AllowVolumeExpansion != nil && *sc.AllowVolumeExpansion {
+				allowed = false
+				result = &metav1.Status{
+					Reason: volumeExpansionErrorMessage,
+				}
+			}
+		}
+		// Migration parameters check for csi.vsphere.vmware.com provisioner
+		if allowed && sc.Provisioner == "csi.vsphere.vmware.com" {
+			for param := range sc.Parameters {
+				if unSupportedParameters.Has(param) {
+					allowed = false
+					result = &metav1.Status{
+						Reason: migrationParamErrorMessage,
+					}
+					break
+				}
+			}
+		}
+	}
+	// return AdmissionResponse result
```

Comment:
Can we have an info level log here stating that the storage class was unsuccessfully/successfully validated?

Reply:
added info level log

```
log.Infof("Validation of StorageClass: %q Passed", sc.Name)
```

```
log.Infof("Validation of StorageClass: %q Failed", sc.Name)
```